File: tests/test_crossentropy.py
import torch
import torch.nn.functional as F
import numpy as np
import sys

# 1. Setup deterministic environment
rng = np.random.default_rng(seed=42)
batch_size, vocab_size = 2, 5

# 2. Logits (Requires Grad!)
logits_np = rng.standard_normal((batch_size, vocab_size)).astype(np.float32)
logits = torch.tensor(logits_np, requires_grad=True)

# 3. Create Targets for the 3 Paths
# A. Indexed
targets_indexed_np = np.array([3, 1], dtype=np.int64)
targets_indexed = torch.tensor(targets_indexed_np)

# B. One-Hot
targets_one_hot_np = np.zeros((batch_size, vocab_size), dtype=np.float32)
targets_one_hot_np[0, 3] = 1.0
targets_one_hot_np[1, 1] = 1.0
targets_one_hot = torch.tensor(targets_one_hot_np)

# C. Probs
targets_probs_np = rng.random((batch_size, vocab_size)).astype(np.float32)
targets_probs_np = targets_probs_np / targets_probs_np.sum(axis=-1, keepdims=True)
targets_probs = torch.tensor(targets_probs_np)

# 4. Incoming Gradient (dLoss)
dLoss_np = rng.standard_normal((batch_size,)).astype(np.float32)
dLoss = torch.tensor(dLoss_np)

# 5. Forward and Backward Passes
# A. Indexed
loss_indexed = F.cross_entropy(logits, targets_indexed, reduction='none')
loss_indexed.backward(dLoss, retain_graph=True)
dX_indexed = logits.grad.clone()
logits.grad.zero_()

# B. One-Hot
loss_one_hot = F.cross_entropy(logits, targets_one_hot, reduction='none')
loss_one_hot.backward(dLoss, retain_graph=True)
dX_one_hot = logits.grad.clone()
logits.grad.zero_()

# C. Probs
loss_probs = F.cross_entropy(logits, targets_probs, reduction='none')
loss_probs.backward(dLoss, retain_graph=True)
dX_probs = logits.grad.clone()
logits.grad.zero_()

# 6. Export to C Header
with open("test_crossentropy.h", "w") as f:
    outputs = [
        f"f32 {name}[] = {{ { np.array2string(param.detach().numpy().reshape(-1), separator=', ',threshold=sys.maxsize)[1:-1] } }};" 
        for name, param in [
            ("logits_np", logits), ("dLoss_np", dLoss),
            ("targets_one_hot_np", targets_one_hot), ("targets_probs_np", targets_probs),
            ("loss_indexed_target", loss_indexed), ("loss_one_hot_target", loss_one_hot), ("loss_probs_target", loss_probs),
            ("dX_indexed_target", dX_indexed), ("dX_one_hot_target", dX_one_hot), ("dX_probs_target", dX_probs)
        ]
    ]
    # Indexed targets are unsigned ints
    outputs.append(f"u32 targets_indexed_np[] = {{ { np.array2string(targets_indexed_np, separator=', ',threshold=sys.maxsize)[1:-1] } }};")

    f.write(f"#ifndef TEST_CROSSENTROPY_H\n#define TEST_CROSSENTROPY_H\n#include \"type_alias.h\"\n#define VOCAB_SIZE {vocab_size}\n#define BATCH_SIZE {batch_size}\n")
    for line in outputs:
        f.write(line + "\n")
    f.write("#endif\n")

print("Generated test_crossentropy.h successfully!")

File: tests/test_crossentropy.c
#include "test_crossentropy.h"
#include "tensor.h"
#include "var.h"
#include <math.h>
#include <stdio.h>
#include <string.h>

#define EPS 1e-4

bool verify_tensor(Tensor out, f32 *target, const char *test_name) {
  if (!out) {
    printf("[FAIL] %s: Output is null.\n", test_name);
    return false;
  }
  for (u32 i = 0; i < out->data->nelements; ++i) {
    if (fabsf(out->data->data[i] - target[i]) > EPS) {
      printf("[FAIL] %s: Mismatch at index %u (Expected: %f, Got: %f)\n",
             test_name, i, target[i], out->data->data[i]);
      return false;
    }
  }
  printf("[PASS] %s\n", test_name);
  return true;
}

// Assumes definitions for var_crossentropy_indexed, var_crossentropy_one_hot,
// var_crossentropy_probs
int main() {
  u32 shape2D[] = {BATCH_SIZE, VOCAB_SIZE};
  u32 shape1D[] = {BATCH_SIZE};

  auto dLoss = tensor_new(1, shape1D);
  memcpy(dLoss->data->data, dLoss_np, BATCH_SIZE * sizeof(f32));

  // ---------------------------------------------------------
  // 1. ONE-HOT PATH
  // ---------------------------------------------------------
  auto logits_oh = track_replace_untracked(tensor_new(2, shape2D));
  memcpy(logits_oh->data->data, logits_np,
         BATCH_SIZE * VOCAB_SIZE * sizeof(f32));

  auto targets_oh = tensor_new(2, shape2D);
  memcpy(targets_oh->data->data, targets_one_hot_np,
         BATCH_SIZE * VOCAB_SIZE * sizeof(f32));

  auto loss_oh = var_cross_entropy_one_hot(logits_oh, targets_oh, 1);
  verify_tensor(loss_oh, loss_one_hot_target, "One-Hot Forward");

  var_backward_with_grad(loss_oh, dLoss);
  verify_tensor(var_grad(logits_oh), dX_one_hot_target, "One-Hot Backward");

  // ---------------------------------------------------------
  // 2. PROBS PATH
  // ---------------------------------------------------------
  auto logits_pr = track_replace_untracked(tensor_new(2, shape2D));
  memcpy(logits_pr->data->data, logits_np,
         BATCH_SIZE * VOCAB_SIZE * sizeof(f32));

  auto targets_pr = tensor_new(2, shape2D);
  memcpy(targets_pr->data->data, targets_probs_np,
         BATCH_SIZE * VOCAB_SIZE * sizeof(f32));

  auto loss_pr = var_cross_entropy_probs(logits_pr, targets_pr, 1);
  verify_tensor(loss_pr, loss_probs_target, "Probs Forward");

  var_backward_with_grad(loss_pr, dLoss);
  verify_tensor(var_grad(logits_pr), dX_probs_target, "Probs Backward");

  // ---------------------------------------------------------
  // 3. INDEXED PATH
  // ---------------------------------------------------------
  auto logits_idx = track_replace_untracked(tensor_new(2, shape2D));
  memcpy(logits_idx->data->data, logits_np,
         BATCH_SIZE * VOCAB_SIZE * sizeof(f32));

  auto targets_idx = tensor_new(1, shape1D);
  for (u32 i = 0; i < BATCH_SIZE; ++i) {
    targets_idx->data->data[i] = (f32)
        targets_indexed_np[i]; // Store as float if tensor structure requires
                               // it, or handle integer tensors accordingly
  }

  auto loss_idx = var_cross_entropy_indexed(logits_idx, targets_idx, 1);
  verify_tensor(loss_idx, loss_indexed_target, "Indexed Forward");

#ifndef UNDEF_WHEN_INDEXED_BP_IS_IMPLEMENTED
  printf("[SKIP] Indexed Backward (Not implemented yet)\n");
#else
  var_backward_with_grad(loss_idx, dLoss);
  verify_tensor(var_grad(logits_idx), dX_indexed_target, "Indexed Backward");
#endif

  return 0;
}
